File: MinCostForTickets.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    /// Time Complexity : O(W) W = 365
    // Space Complexity :O(W) 
    // Did this code successfully run on Leetcode :Yes
    // Any problem you faced while coding this :  No 
    public class MinCostForTickets
    {
        int[] costs;
        int[] memo;
        HashSet<int> set;
        public int MincostTickets(int[] days, int[] costs)
        {
            this.costs = costs;
            memo = new int[366];
            set = new HashSet<int>();
            for (int i = 0; i < days.Length; i++)
            {
                set.Add(days[i]);
            }

            return dp(1);
        }

        public int dp(int i)
        {
            if (i > 365) return 0;
            if (memo[i] != 0)
            {
                return memo[i];
            }
            int ans;
            if (set.Contains(i))
            {
                ans = Math.Min((dp(i + 1) + costs[0]), (dp(i + 7) + costs[1]));
                ans = Math.Min(ans, (dp(i + 30) + costs[2]));
            }
            else
            {
                ans = dp(i + 1);
            }
            memo[i] = ans;
            return ans;
        }

    }
}

File: minCostForTickets.java
// Time Complexity : O(max(days))
// Space Complexity : O(max(days))
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
/*
 * 1. Create a hashset of all the travel days. It is used to check if a day is a travel day or not.
 * 2. Create a memo array to store the minimum cost for each day.
 * 3. If the current day is not a travel day, then move to the next day.
 * 4. If the current day is a travel day, then calculate the cost for 1 day, 7 days and 30 days and return the minimum of the three.
 */

import java.util.Arrays;
import java.util.HashSet;

class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        HashSet<Integer> travelDays = new HashSet<>();
        int[] memo = new int[days[days.length-1]+1];
        Arrays.fill(memo, -1);

        for(int day : days){
            travelDays.add(day);
        }

        return helper(days, costs, days[0], travelDays, memo);
    }

    private int helper(int[] days, int[] costs, int currDay, HashSet<Integer> travelDays, int[] memo){
        if(currDay > days[days.length-1]){
            return 0;
        }

        if(memo[currDay] != -1){
            return memo[currDay];
        }

        if(!travelDays.contains(currDay)){
            return memo[currDay] = helper(days, costs, currDay+1, travelDays, memo);
        }

        int cost1 = costs[0] + helper(days, costs, currDay+1, travelDays, memo);
        int cost2 = costs[1] + helper(days, costs, currDay+7, travelDays, memo);
        int cost3 = costs[2] + helper(days, costs, currDay+30, travelDays, memo);

        return memo[currDay] = Math.min(cost1, Math.min(cost2, cost3));
    }
}